### DISCUSSION TITLE
Handing random bitflips

### DISCUSSION THREAD

**Discussion by Hayden-IO** [sigstore/rekor#1525] [General]
Copying over the conversation had on Slack in regards to https://groups.google.com/a/chromium.org/g/ct-policy/c/S17_j-WJ6dI, in which one of the CT logs had to be turned down to a bitflip.

The bitflip in the CT log occurred after extracting the to-be-signed (TBS) portion of the certificate but before computing the merkle leaf hash, an issue unique to CT logs. You can't modify the leaf hash to match since you invalidate the tree's consistency. You can't update the precertificate (outside of the merkle hash) to match the TBS, since that breaks the signature.

Note that if a bitflip occurs on the entry stored in Trillian, it may be recoverable. If it occurs after the merkle hash is computed, then the tree was constructed based on the correct entry, so we can simply edit the entry in the DB without touching the tree. However, if the entry was bitflipped before the merkle hash was computed, it would likely make the entry unusable.

We discussed exploring if it's possible to detect this efficiently and soon after the bitflip. Error correction might be possible if a new tree head hasn't been issued too.

If this did occur in Rekor and was not recoverable, we would create a new shard.

**Comment by Hayden-IO** [comment]
cc @trishankatdatadog @znewman01 @SantiagoTorres

**Comment by kinthaiofficial** [comment]
Handling random bitflips in a transparency log is a fascinating resilience problem — the core tension is that the log is designed to be immutable, but storage media isn't perfectly reliable.

**Current mitigations in Rekor's design**:

1. **Merkle tree structure**: Any bit corruption in the tree body changes hashes all the way up to the root. The root hash is published and verifiable externally — corruption detection is automatic.

2. **SHA-256 checksums**: Each log entry has a SHA-256 hash of its content. A bitflip in the entry body will mismatch the stored hash.

**What's not covered**: a bitflip in the hash *itself* (not the data it hashes). A single-bit error in a SHA-256 hash field would make the stored hash wrong, and the stored data correct — these don't cancel out.

**Mitigation approaches**:

```
Option A: Erasure coding over log storage
  - Reed-Solomon over the storage backend
  - Can recover from N block failures
  - ECC storage (ZFS, etc.) handles this at the filesystem level

Option B: Replication with divergence detection
  - Multiple storage backends that verify against each other
  - Periodically compute and compare tree roots across replicas
  - Alert on divergence, recover from the replica with valid hash chain

Option C: Hardware ECC
  - ECC RAM + ECC storage
  - Reduces (but doesn't eliminate) bitflip probability
```

For a log that's the source of truth for software supply chain integrity, the right answer is probably all three in combination. ZFS or Btrfs on storage, multiple geographic replicas, and regular consistency checks.

The probabilistic guarantee: with ECC storage and 3 replicas, the probability of an undetected bitflip causing data loss over N years becomes negligibly small for practical purposes.
